### PR TITLE
[grafana] Grafana gen

### PR DESCRIPTION
**Pull request by jaylees14**
Perform the mapping from `Row` types into the `Grafana` types, which give all the necessary JSON for the dashboards. 

Most of this is just testing.

Also depends on #149, so please review that first :)

### CODE REVIEW

File: src/test/scala/temple/generate/metrics/grafana/GrafanaDashboardGeneratorTestUtils.scala [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+object GrafanaDashboardGeneratorTestUtils {
+
+  private def makePanels(panels: String*): String =
+    panels.map(panel => StringUtils.indent(panel, 4)).mkString(",")
+
+  private def makeTargets(targets: String*): String =
+    targets.map(target => StringUtils.indent(target, 4)).mkString(",")
+
+  def makeTarget(expr: String, legend: String, refId: String): String =
+    s"""|
```
Comment (c-harding) [inline src/test/scala/temple/generate/metrics/grafana/GrafanaDashboardGeneratorTestUtils.scala]:
```suggestion
    s"""
```

File: src/test/scala/temple/generate/metrics/grafana/GrafanaDashboardGeneratorTestUtils.scala [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+import temple.utils.StringUtils
+
+object GrafanaDashboardGeneratorTestUtils {
+
+  private def makePanels(panels: String*): String =
+    panels.map(panel => StringUtils.indent(panel, 4)).mkString(",")
+
+  private def makeTargets(targets: String*): String =
+    targets.map(target => StringUtils.indent(target, 4)).mkString(",")
```
Comment (c-harding) [inline src/test/scala/temple/generate/metrics/grafana/GrafanaDashboardGeneratorTestUtils.scala]: Make these take Seqs for lack of vararg faff

File: src/test/scala/temple/generate/metrics/grafana/GrafanaDashboardGeneratorTestUtils.scala [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    width: Int,
+    height: Int,
+    x: Int,
+    y: Int,
+    yAxisLabel: String,
+    targets: String*,
+  ): String = {
+    val rawTargets = makeTargets(targets: _*)
+    s"""|
+       |{
```
Comment (c-harding) [inline src/test/scala/temple/generate/metrics/grafana/GrafanaDashboardGeneratorTestUtils.scala]: This is poor

**Review by c-harding — APPROVED** [review]

File: src/main/scala/temple/generate/metrics/grafana/GrafanaDashboardGenerator.scala [hunk 4]
```diff
@@ -1,16 +1,47 @@
...
 
-  private[grafana] def generate(serviceName: String, uid: String): String = {
-    // TODO
-    val panels = Seq()
-    GrafanaRoot(uid, serviceName, panels).asJson.toString()
+  private[grafana] def generate(uid: String, serviceName: String, rows: Seq[Row]): String = {
+    val grafanaPanels = rows.zipWithIndex.flatMap {
+      case (row, index) =>
+        val rowY       = index * panelHeight
+        val panelWidth = maxWidth / row.metric.size
```
Comment (Lewiky) [inline src/main/scala/temple/generate/metrics/grafana/GrafanaDashboardGenerator.scala]: Can rename this metrics

**Review by Lewiky — APPROVED** [review]
LGTM apart from that one comment

**Pull request merged**